subsys/net/ip/tcp2.c

```diff
@@ -925,6 +925,9 @@ static int tcp_send_data(struct tcp *conn)
 	len = MIN3(conn->send_data_total - conn->unacked_len,
 		   conn->send_win - conn->unacked_len,
 		   conn_mss(conn));
+	if (len == 0) {
+		goto out;
```

Comment:
Hmm, shouldn't we return <0 value in this case, now 0 value means in `tcp_resend_data()` that it will increase the retry counter but in this case there is nothing to send.

Reply:
I agree, plus I'd add a log message here, so that it's visible how often this occurs.

Reply:
Is this OK?
```suggestion
		NET_ERR("conn: %p no data to send", conn);
		ret = -ENODATA;
		goto out;
```

Reply:
Fine by me

Reply:
IMHO `NET_DBG()` is enough here, this is not a very serious issue. Is there an easy way to trigger this condition, how are you testing this?

Reply:
32 TCP clients running on host send data to 32 TCP echo server threads running on native posix, same test as in #32564.

Reply:
2 TCP clients send data to 2 TCP echo server threads can trigger this condition. 
I add k_uptime_get() in log message to see how often this condition occurs for this test.
```
E: conn: 0x567eaef0 no data to send, k_uptime_get()=5400
E: conn: 0x567eaef0 no data to send, k_uptime_get()=7600
E: conn: 0x567eb414 no data to send, k_uptime_get()=8550
E: conn: 0x567eb414 no data to send, k_uptime_get()=20000
E: conn: 0x567eaef0 no data to send, k_uptime_get()=121700
E: conn: 0x567eaef0 no data to send, k_uptime_get()=152050
E: conn: 0x567eb414 no data to send, k_uptime_get()=243450
```